#	Handling Errors in Angular

## Types of Errors

-	Unexpected Errors

	-	Server is offline
	-	N/w is  down
	-	Unhandled Exceptions
	
-	Expected Errors
	
	-	Not found - 404 Errors
		-	Invalid resource url
	-	Bad Request - 400 Errors
		-	Client sending bad request to server

##	Handling Unexpected Errors

-	Using Subscribe
	
	
			ngOnInit(){
				this.service().getPosts()
				.subscribe(response => {
				  this.data = response.json();
				}, error => {
					alert('An Unexpected Error');
					console.log();
				})
			}
			
##	Handling Expected Errors

			ngOnInit() {
				
				this.service.getPosts()
				.subscribe(
					(response: Response) => {
						this.data= response.json();
					},(error: Response) => {
						if(error.status === 404){
							alert('This post has already been deleted');
						} else {
							alert('An Expected Error');
							console.log();
						}			
					} );
			
			}
			
##	Using error.json() 

-	We can use error.json() method similar to response.json() method
-	We should not write error handling logic in the component and needs to move to the service layer to achieve separation of concerns


		form: NgForm = new NgForm();
		
		ngOnInit() {
			this.service.createPost(post)
				.subscribe(
					(response: Response) => {
						this.data = response.json();
					},
					(error: Response) => {
						if(error.status === 400){
							this.form.errors = error.json();
						} else {
							alert('An unexpected error');
							console.log('Error');
						} 
					});
		}



##	Throwing Application Specific Error

-	Angular Http module returns Observable as an return type of Http call
-	Observable is an class from reactive framework call rxjs 
-	In-order to create a new instance of Observable we need to import Observable from rxjs module
-	Observable provides a catch method or operator to catch the error or exceptions from API Http calls

	-	import { Observable } from 'rxjs';


			
					@Injectable
					export class PostHttpService{
					
						constructor(private _http: Http) {
						}
						
						deletePost(id: number) {
							this._http.delete(apiUrl+'/'+id)
							.do()
							.map((response: Response) => {})
							.catch((error: Error) => {
								if(error instanceOf NotFoundError) {
									Observable.throw(new NotFoundError(error));
								} else {
									Observable.throw(new ApiError(error));
								}
							});
						}
					}
					
					
					
					export class ApiError {
						constructor (public apiError?: any){
						}
					}
					
					
					export class NotFoundError extends ApiError {
						
					}





## Handling Bad Request Errors


			export class BadRequestError extends ApiError {
			
			}
		
			@Injectable
			export class PostHttpService{
				
				constructor(private _http: Http) {
				}
				
				
				createPost(post : Post) {
					
					this._http.post(apiUrl, Json.stringfy(post))
					.catch((error: ApiError) => {
						if (error instanceOf  BadRequestError) {
							this.form.setErrors(new BadRequestError(error.apiError))
						} else {
							alert('error in creating post')
						}
					});
				
				} 
				
				deletePost(id: number) {
					this._http.delete(apiUrl+'/'+id)
					.do()
					.map((response: Response) => {})
					.catch((error: Error) => {
						if(error instanceOf NotFoundError) {
							Observable.throw(new NotFoundError(error));
						} else {
							Observable.throw(new ApiError(error));
							console.log('Some thing very Bad error');
						}
					});
				}
				
				
				
				
			
			}


## Handling Error Global

-	Angular provides a built in class call ErrorHandler to handle the errors globally
-	Angular by default uses this call to log the error to console
-	We can implement this interface and provide custom implementation to handle error globally


	-	Register with AppModule to use ApiErrorHandler as default Error handler class
				
				@NgModule({
					providers: [
						{ provide : ErrorHandler, useClass : AppErrorHandler }
					]
				})
				export class AppModule {}

				export class ApiErrorHandler implements ErrorHandler {
				
					handleError(error: any) {
						
					}
				
				}
						

				@Injectable
					export class PostHttpService{
						
						constructor(private _http: Http) {
						}
						
						
						createPost(post : Post) {
							
							this._http.post(apiUrl, JSON.stringfy(post))
							.catch((error: ApiError) => {
								if (error instanceOf  BadRequestError) {
									this.form.setErrors(new BadRequestError(error.apiError))
								} else throw error;
							});
						}
						
						deletePost(id: number) {
							this._http.delete(apiUrl+'/'+id)
							.do()
							.map((response: Response) => {})
							.catch((error: Error) => {
								if(error instanceOf NotFoundError) {
									Observable.throw(new NotFoundError(error));
								} else throw error;
							});
				
						getPosts() {
							this._http.get(apiUrl)
							.do(console.log('working fine'))
							.map((response: Response) => this.data = response.json());
						}
						
					} 


## Observable

-	do(), map(), catch(),subscribe() methods are operators on the Observable class used for asynchronous operation
-	map() operator is used to transform object of one type to another 

	-	subscribe(
			(response : Response) => {
			
			},
			(error: Response) => {
			
			
		});
		
	-	catch((error:Response) => {
	
	
		})
		
	-	map((response: Response) => {
			
		
		});
		
	-	new Promise((resolve, reject) => {
			setTimeout(() => {
							
				if(control.value === 'bharath'){
					resolve({isUnique: true});
				}
				else {
					resolve(null);
				}
			
			
			},3000);
			
		});


##	Observables vs Promises


-	With Observable nothing happens until we subscribe to that ie API call is not going to happen 
-	Observables are lazy and promises are eager
-	We can convert and Observable to promise but not recommended using toPromise() operator on Observable instance

		
		this._http.get()
		.map((response:Response) => this.data = response.json())
		.toPromise()
		.catch();

-	promises dont have subscribe() method
-	promises have only two methods 

	-	then(): for handling success response
	- 	catch(): for handling error response
-	promises are eager, we dont need to call then() method to make an API call
-	Observablesa are lazy nothing happens until we subscribe to them


##	Why Observables are lazy

-	Observables are lazy because of its operators 
-	Observable operators are very powerful to perform specific operations
-	Observable operators allows to implement certain features with only one of code 
-	retry(3) 

	-	if call to API fails observable retry(3) method is going to make API 3 times 
	-	traditionally retry() requires lot of code but with observable retry method its just a one line of code
	-	This is called Reactive Programming
		-	performing complex operations with one line of code
		-	rxjs is library for using Reactive Programming allowing us to write code in reactive style
		-	Reactive Programming allows some feature that is far more complex to implement the traditional way
		
		
-	Since observable is going to work only when we subscribe to it, we can chain multiple operators together like map(), catch(), do(), retry(3) etc






















		